newIDE/app/src/AssetStore/PrivateAssets/AuthorizedAssetImage.js

```diff
@@ -0,0 +1,105 @@
+// @flow
+import * as React from 'react';
+import PrivateAssetsAuthorizationContext from './PrivateAssetsAuthorizationContext';
+import PlaceholderLoader from '../../UI/PlaceholderLoader';
+import { createAuthorizedUrl } from '../../Utils/GDevelopServices/Shop';
+
+type Props = {|
+  url: string,
+  style?: Object,
+  alt: string,
+  onError?: () => void,
+  onLoad?: (e: any) => void,
+  hideLoader?: boolean,
+|};
+
+const AuthorizedAssetImage = ({
+  url,
+  style,
+  alt,
+  onError,
+  onLoad,
+  hideLoader,
+}: Props) => {
+  const [authorizedUrl, setAuthorizedUrl] = React.useState(null);
+  const [isImageLoaded, setIsImageLoaded] = React.useState(false);
+  const { authorizationToken, updateAuthorizationToken } = React.useContext(
+    PrivateAssetsAuthorizationContext
+  );
+  const [tries, setTries] = React.useState(0);
+
+  React.useEffect(
+    () => {
+      // Avoid trigerring a new request if the image is already loaded.
+      // This can happen if it was loaded with a previous token, and
+      // A new token is fetched for another image.
+      if (isImageLoaded) return;
+
+      if (!authorizationToken) {
+        updateAuthorizationToken();
+      } else {
+        setAuthorizedUrl(createAuthorizedUrl(url, authorizationToken));
+      }
+    },
+    [authorizationToken, updateAuthorizationToken, url, isImageLoaded]
+  );
+
+  // React.useEffect(
+  //   () => {
+  //     // Ensure we reset the loaded status when the URL changes.
+  //     // Useful when used in the AnimationPreview.
+  //     if (url) {
+  //       setIsImageLoaded(false);
+  //     }
+  //   },
+  //   [url]
+  // );
+
+  const onFetchingError = () => {
+    console.warn('Error while fetching authorized image');
+    if (tries >= 3) {
+      if (onError) onError();
+      return;
+    }
+    if (!hideLoader) {
+      setIsImageLoaded(false);
+    }
+    // If the image is not authorized, fetch a new authorization token.
+    console.info('Error while fetching image, fetching a new token...');
+    updateAuthorizationToken();
+    setTries(tries + 1);
+  };
+
+  const onImageLoaded = (e: any) => {
+    if (!hideLoader) {
+      setIsImageLoaded(true);
```

Comment:
Should you set it to true even if the loader should be hidden?

Reply:
Since it impacts the authorization token update logic

Reply:
yep, I overengineered the logic here